--- /Users/emonson/Programming/VTK_git/VTK/Charts/vtkPlotPoints.h	2010-06-30 10:56:13.000000000 -0400
+++ vtkMyPlotPoints.h	2010-06-30 09:45:39.000000000 -0400
@@ -1,7 +1,7 @@
 /*=========================================================================
 
   Program:   Visualization Toolkit
-  Module:    vtkPlotPoints.h
+  Module:    vtkMyPlotPoints.h
 
   Copyright (c) Ken Martin, Will Schroeder, Bill Lorensen
   All rights reserved.
@@ -13,32 +13,39 @@
 
 =========================================================================*/
 
-// .NAME vtkPlotPoints - Class for drawing an XY plot given two columns from a
+// .NAME vtkMyPlotPoints - Class for drawing an XY plot given two columns from a
 // vtkTable.
 //
 // .SECTION Description
 //
 
-#ifndef __vtkPlotPoints_h
-#define __vtkPlotPoints_h
+#ifndef __vtkMyPlotPoints_h
+#define __vtkMyPlotPoints_h
 
 #include "vtkPlot.h"
+#include "vtkSmartPointer.h"
+// #include "vtkVector.h"
 
 class vtkContext2D;
 class vtkTable;
 class vtkPoints2D;
 class vtkStdString;
 class vtkImageData;
+class vtkMatrix4x4;
+class vtkImageReslice;
+class vtkLookupTable;
+class vtkImageMapToColors;
+class vtkVector3f;
 
-class VTK_CHARTS_EXPORT vtkPlotPoints : public vtkPlot
+class VTK_CHARTS_EXPORT vtkMyPlotPoints : public vtkPlot
 {
 public:
-  vtkTypeMacro(vtkPlotPoints, vtkPlot);
+  vtkTypeMacro(vtkMyPlotPoints, vtkPlot);
   virtual void PrintSelf(ostream &os, vtkIndent indent);
 
   // Description:
   // Creates a 2D Chart object.
-  static vtkPlotPoints *New();
+  static vtkMyPlotPoints *New();
 
   // Description:
   // Perform any updates to the item that may be necessary before rendering.
@@ -55,17 +62,24 @@
   // plot items symbol/mark/line drawn. A rect is supplied with the lower left
   // corner of the rect (elements 0 and 1) and with width x height (elements 2
   // and 3). The plot can choose how to fill the space supplied.
-  virtual bool PaintLegend(vtkContext2D *painter, float rect[4], int legendIndex);
+  virtual bool PaintLegend(vtkContext2D *painter, float rect[4]);
 
   // Description:
   // Get the bounds for this mapper as (Xmin,Xmax,Ymin,Ymax,Zmin,Zmax).
   virtual void GetBounds(double bounds[4]);
 
+  // Description
+  // ImageData associated with plot, which the tooltip in the chartXY will get
+  // a slice of to display when hovering over points (needs to be just 2d)
+  virtual void SetTooltipImageStack(vtkImageData*);
+  virtual int GetNumberOfImages();
+  virtual vtkImageData* GetImageAtIndex(int imageId);
+  
+  
 //BTX
   // Description:
   // Function to query a plot for the nearest point to the specified coordinate.
-  // Returns the index of the data series with which the point is associated or 
-  // -1.
+  // This version returns the index of the point as a float in the .Z() vector3f coord.
   virtual int GetNearestPoint(const vtkVector2f& point,
                                const vtkVector2f& tolerance,
                                vtkVector2f* location);
@@ -75,6 +89,11 @@
   virtual bool SelectPoints(const vtkVector2f& min, const vtkVector2f& max);
 
   // Description:
+  // Set the list of ids for highlighting (back from outside plot).
+  virtual void SetHighlightSelection(vtkIdTypeArray *id);
+  vtkGetObjectMacro(HighlightSelection, vtkIdTypeArray);
+
+  // Description:
   // Enum containing various marker styles that can be used in a plot.
   enum {
     NONE = 0,
@@ -94,8 +113,8 @@
 
 //BTX
 protected:
-  vtkPlotPoints();
-  ~vtkPlotPoints();
+  vtkMyPlotPoints();
+  ~vtkMyPlotPoints();
 
   // Description:
   // Generate the requested symbol for the plot
@@ -126,8 +145,8 @@
 
   // Description:
   // Sorted points, used when searching for the nearest point.
-  class VectorPIMPL;
-  VectorPIMPL* Sorted;
+  class VectorPIMPL3;
+  VectorPIMPL3* Sorted;
 
   // Description:
   // An array containing the indices of all the "bad points", meaning any x, y
@@ -137,20 +156,35 @@
   // Description:
   // The point cache is marked dirty until it has been initialized.
   vtkTimeStamp BuildTime;
+  
+  // Description
+  // ImageData associated with plot, which the tooltip in the chartXY will get
+  // a slice of to display when hovering over points
+  vtkImageData* TooltipImageStack;
+  int NumImages;
 
   // Description:
   // The marker style that should be used
   int MarkerStyle;
   vtkImageData* Marker;
-  vtkImageData* HighlightMarker;
+  vtkImageData* SelectionMarker;
+
+  // Description:
+  // Selected indices coming back from outside the chart this plot is associated with.
+  vtkIdTypeArray *HighlightSelection;
 
   bool LogX, LogY;
+  
+  vtkSmartPointer<vtkMatrix4x4> resliceAxes;
+  vtkSmartPointer<vtkImageReslice> reslice;
+  vtkSmartPointer<vtkLookupTable> lut;
+  vtkSmartPointer<vtkImageMapToColors> color;
 
 private:
-  vtkPlotPoints(const vtkPlotPoints &); // Not implemented.
-  void operator=(const vtkPlotPoints &); // Not implemented.
+  vtkMyPlotPoints(const vtkMyPlotPoints &); // Not implemented.
+  void operator=(const vtkMyPlotPoints &); // Not implemented.
 
 //ETX
 };
 
-#endif //__vtkPlotPoints_h
+#endif //__vtkMyPlotPoints_h
